channels/rdpemsc/server/mouse_cursor_main.c

```diff
@@ -0,0 +1,727 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Mouse Cursor Virtual Channel Extension
+ *
+ * Copyright 2023 Pascal Nowack <Pascal.Nowack@gmx.de>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <freerdp/config.h>
+
+#include <freerdp/freerdp.h>
+#include <freerdp/channels/log.h>
+#include <freerdp/server/rdpemsc.h>
+
+#define TAG CHANNELS_TAG("rdpemsc.server")
+
+typedef enum
+{
+	MOUSE_CURSOR_INITIAL,
+	MOUSE_CURSOR_OPENED,
+} eMouseCursorChannelState;
+
+typedef struct
+{
+	MouseCursorServerContext context;
+
+	HANDLE stopEvent;
+
+	HANDLE thread;
+	void* mouse_cursor_channel;
+
+	DWORD SessionId;
+
+	BOOL isOpened;
+	BOOL externalThread;
+
+	/* Channel state */
+	eMouseCursorChannelState state;
+
+	wStream* buffer;
+} mouse_cursor_server;
+
+static UINT mouse_cursor_server_initialize(MouseCursorServerContext* context, BOOL externalThread)
+{
+	UINT error = CHANNEL_RC_OK;
+	mouse_cursor_server* mouse_cursor = (mouse_cursor_server*)context;
+
+	WINPR_ASSERT(mouse_cursor);
+
+	if (mouse_cursor->isOpened)
+	{
+		WLog_WARN(TAG, "Application error: Mouse Cursor channel already initialized, "
+		               "calling in this state is not possible!");
+		return ERROR_INVALID_STATE;
+	}
+
+	mouse_cursor->externalThread = externalThread;
+
+	return error;
+}
+
+static UINT mouse_cursor_server_open_channel(mouse_cursor_server* mouse_cursor)
+{
+	MouseCursorServerContext* context = &mouse_cursor->context;
+	DWORD Error = ERROR_SUCCESS;
+	HANDLE hEvent;
+	DWORD BytesReturned = 0;
+	PULONG pSessionId = NULL;
+	UINT32 channelId;
+	BOOL status = TRUE;
+
+	WINPR_ASSERT(mouse_cursor);
+
+	if (WTSQuerySessionInformationA(mouse_cursor->context.vcm, WTS_CURRENT_SESSION, WTSSessionId,
+	                                (LPSTR*)&pSessionId, &BytesReturned) == FALSE)
+	{
+		WLog_ERR(TAG, "WTSQuerySessionInformationA failed!");
+		return ERROR_INTERNAL_ERROR;
+	}
+
+	mouse_cursor->SessionId = (DWORD)*pSessionId;
+	WTSFreeMemory(pSessionId);
+	hEvent = WTSVirtualChannelManagerGetEventHandle(mouse_cursor->context.vcm);
+
+	if (WaitForSingleObject(hEvent, 1000) == WAIT_FAILED)
```

Comment:
What is the reason to use a timeout here (and a magic number at that)?

Reply:
I actually copied that part from the other DVCs, not sure the overall origin (whether it was audin or some other DVC)

Reply:
ok, I´m asking because with `FreeRDP 1` we had issues with events (e.g. not all relevant things were notified, so the hackish way was to use a polling loop or timeouts)

Reply:
not sure about the origin, but it is indeed unneeded, as the event handle for the channel queue is created, when the VCM is created. I removed it in the last push, it is unneeded.